dev/run-tests.py

```diff
@@ -0,0 +1,417 @@
+            " ".join(sbt_maven_profile_args + ["-Phive", "-Phive-thriftserver"])
+
+    # if we only have changes in SQL build a custom test string
+    if "SQL" in test_suite and "CORE" not in test_suite:
+        sbt_maven_test_args = ["catalyst/test",
+                               "sql/test",
+                               "hive/test", 
+                               "hive-thriftserver/test",
+                               "mllib/test"]
+    else:
+        sbt_maven_test_args = ["test"]
+
+    # get the latest sbt maven profile arguments
+    sbt_maven_profile_args = os.environ.get(SBT_MAVEN_PROFILE_ARGS_ENV).split()
+
+    print "[info] Running Spark tests with these arguments:",
+    print " ".join(sbt_maven_profile_args), 
+    print " ".join(sbt_maven_test_args)
+
+    if AMPLAB_JENKINS_BUILD_TOOL == "maven":
+        exec_maven(["test"] + sbt_maven_profile_args + ["--fail-at-end"])
+    else:
+        exec_sbt(sbt_maven_profile_args + sbt_maven_test_args)
+
+
+def run_python_tests(test_suite=[]):
+    set_title_and_block("Running PySpark tests", "BLOCK_PYSPARK_UNIT_TESTS")
+    
+    # Add path for Python3 in Jenkins if we're calling from a Jenkins machine
+    if AMPLAB_JENKINS:
+        os.environ["PATH"] = os.environ.get("PATH")+":/home/anaconda/envs/py3k/bin"
+
+    run_cmd(["./python/run-tests"])
+
+
+def run_sparkr_tests(test_suite=[]):
+    set_title_and_block("Running SparkR tests", "BLOCK_SPARKR_UNIT_TESTS")
+
+    if which("R"):
+        run_cmd(["./R/install-dev.sh"])
+        run_cmd(["./R/run-tests.sh"])
+    else:
+        print "Ignoring SparkR tests as R was not found in PATH"
+
+if __name__ == "__main__":
+    # Ensure the user home directory (HOME) is valid and is an absolute directory
+    if not USER_HOME_DIR or not os.path.isabs(USER_HOME_DIR):
+        print "[error] Cannot determine your home directory as an absolute path;",
+        print "ensure the $HOME environment variable is set properly."
+        sys.exit(1)
+
+    os.chdir(SPARK_PROJ_ROOT)
+
+    rm_r("./work")
+    rm_r(os.path.join(USER_HOME_DIR, ".ivy2/local/org.apache.spark"))
+    rm_r(os.path.join(USER_HOME_DIR, ".ivy2/cache/org.apache.spark"))
+
+    error_codes = get_error_codes("./dev/run-tests-codes.sh")
+
+    os.environ["CURRENT_BLOCK"] = error_codes["BLOCK_GENERAL"]
+
+    set_sbt_maven_profile_args()
+
+    java_exe = determine_java_executable()
+
+    if not java_exe:
+        print "[error] Cannot find a version of `java` on the system; please", 
+        print "install one and retry."
+        sys.exit(2)
+
+    java_version = determine_java_version(java_exe)
+
+    if java_version.minor < 8:
+        print "[warn] Java 8 tests will not run because JDK version is < 1.8."
+
+    test_suite = determine_test_suite()
+
+    run_apache_rat_checks()
+    
+    run_scala_style_checks()
+
+    run_python_style_checks()
+
+    build_apache_spark()
+
+    detect_binary_inop_with_mima()
+
+    run_scala_tests(test_suite)
```

Comment:
Btw - @brennonyork one other small enhancement you could add is to also support "ml-only" and "streaming-only" based on the git diff. I realize that is a slight scope creep in this (since it isn't supported in the old scripts) but it would be such a high value change for our tests (will literally save us hours of developer time every week) that if it's simple to just do it here I think it's worth doing.